### PR TITLE
Valgrind

### PR DESCRIPTION
**Pull request by fennecdjay**
Also a few improvements:
  * test suite now runs local zpm.
  * travis build use containers.
  * `Makefile` now properly uses clang.
  * added `clean` target.

### CODE REVIEW

**Comment by fennecdjay** [comment]
`feea535` fails on OSX valgrind tests.
Showing the log and making valgrind more verbose on failure would ease debugging.

**Comment by desyncr** [comment]
Well this is exactly what we're looking for! To have valgrind to cover us in regard of memory leaks and hidden bugs. So it's all fine!

Now, do you want to try to fix the memory issues in this PR or should I merge this already (haven't done reviewing it yet, still) and we'll catch the memory issues later on?

**Comment by fennecdjay** [comment]
> we'll catch the memory issues later on?
I'd prefer that, but we should not merge before we see the valgrind log on errors.
I'll work on this ASAP.

**Comment by fennecdjay** [comment]
Sorry for the delay.
I think now (with allowed_failures and loggin on errors), it'll be fine.

**Comment by desyncr** [comment]
I'll try this later tonight. Looking fine and it's a great addition. My only nitpick is having to use `$ZPM` in cram tests. Can aliases be used instead?

Regards for `v0.1.0` @fennecdjay https://github.com/zpm-project/zpm-zsh/releases/tag/v0.1.0

**Comment by fennecdjay** [comment]
> Can aliases be used instead?

I'd like that, but never managed to do so 😕 
Any hint appreciated here 😄

**Comment by fennecdjay** [comment]
Now all leaks come from url.h.
Maybe we should fix it (along with for loop declarations) in a fork and make a pull request.
What do you think?
(I'm not even sure where to discuss this 😖 )

----------

EDIT: you can have a look a jwerle/url.h#3.

I tend to think we could ship a refined zpm-project version of this.

**Comment by fennecdjay** [comment]
> Can aliases be used instead?

I've been rechecking cram's doc and issue tracker, and did not find a clue on how to do that.
From what I understand, `cram` launches a new shell on each test file.
All my previous trials failed, so unless you got a way to achive this that you can show me, I think we'll have to stick to the `$ZPM` thing.

**Comment by desyncr** [comment]
@fennecdjay We can contribute to `url.h` project, either by raising an issue or creating a PR (it's always better to start with an issue to check if there is an interest from the other side).

In regard of aliases I couldn't do it right away. I'm OK with `$ZPM` if there is not another option.

**Comment by fennecdjay** [comment]
I created an issue, and a PR would be welcome :tada:.
see  jwerle/url.h#4

**Comment by desyncr** [comment]
@fennecdjay Yay! Awesome!

File: zpm.c [hunk 1]
```diff
@@ -363,15 +364,18 @@ char* plugin_get_hash(char* plugin_name) {
     char* plugin_path = generate_plugin_path(plugin_name);
     char plugin_hash[PATH_MAX];
     char command[PATH_MAX];
+	char git_path[PATH_MAX];
 
-    strcat(plugin_path, "/.git");
-    if (stat(plugin_path, &s) == -1 || !S_ISDIR(s.st_mode)) {
+	strcpy(git_path, plugin_path);
```
Comment (desyncr) [inline zpm.c]: Tabbing issue.

File: zpm.c [hunk 2]
```diff
@@ -363,15 +364,18 @@ char* plugin_get_hash(char* plugin_name) {
...
+    strcat(git_path, "/.git");
+	memset(&s, 0, sizeof(s));
```
Comment (desyncr) [inline zpm.c]: Same here.

**Review by desyncr — APPROVED** [review]
Two small code style things. Ready to merge from my side. Anything else to add?

**Comment by fennecdjay** [comment]
I'll fix this soon. Should we use something like [astyle](), or whichever you prefer, along with a config file to get rid of those annoyances?

**Comment by fennecdjay** [comment]
I think there nothing else to add here. we could merge and move to next steps.

**Comment by desyncr** [comment]
Awesome! 👍  Great addition!

**Pull request merged**